### ISSUE TITLE
set enclosing environment of RCfunctions based on getNimbleFunctionEnvironment()

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#876]
@danielturek is developing a package and when he tries to use one RCfunction within another RCfunction, R CMD check complains like this:

```
* checking R code for possible problems ... NOTE
calculateU_ns: no visible global function definition for ‘nsCrosscorr’
calculateU_ns: no visible global function definition for ‘nsCorr’
```

Apparently R CMD check evaluates the code in the package R files. This evaluation creates a set of functions. For each of these given functions, for each function call in the body of  the function, it looks through the enclosing environments of the given function to try to find the functions that are called.

When an RCfunction is created, nimbleFunction() calls RCfunction() and ultimately the RCfunction is created in a new environment by nfMethodRC$generateFunctionObject(). Because the RCfunction is defined within nimble package code, one of the RCfunction's enclosing environments is the nimble namespace. (There are a couple intervening enclosing environments - those of generateFunctionObject and of nfMethodRCclass.)

E.g., just for RCfunction created in a plain R session:
```
parent.env(parent.env(parent.env(environment(myRCfunction))))
<environment: namespace:nimble>
```

The upshot is that the RCfunction that is created in a package (that is not 'nimble' itself) does not have the package namespace in its enclosing environments. So R CMD check cannot find functions that are from that package namespace. 

I believe the solution is to pass "where" into RCfunction and on into nfMethodRC$generateFunctionObject. There we should do this:

```
environment(ans) <- new.env(parent = where)
```

I think this would be fine to do for RCfunctions created within the nimble package itself as well. But we should check carefully the series of enclosing environments that are currently being set for RCfunctions in the nimble namespace. Currently I see:
```
 parent.env(parent.env(parent.env(environment(nimble:::CAR_calcCmatrix))))
<environment: namespace:nimble>
```

and the proposal would eliminate a couple of these intervening enclosing environments.

Also we might as well have the name of the RCfunction make use of `where` as we do for nimbleFunctions. So we may want to pass in `where` into rcFunLabelMaker.

Notes:
1) because nimbleFunctions with setup code produce generators and there is no run code in the body of the generator function, this issue hasn't arisen for nimbleFunctions. Namely R CMD check never even looks at the $run code.
2) if one wants to see the nimbleFunction reference class definitions in a package namespace, you won't see them with ls('package:foo'). Instead type "foo:::.__" and hit tab and auto-completion should reveal things like `nimble:::.__C__CAR_scalar_RW`

@perrydv @danielturek let's discuss at our next meeting, but we can also iterate here if you like.

**Comment by danielturek** [comment]
I followed most of that, and thanks for the detailed investigation and report @paciorek.

To be clear about exactly where this is at:

1. Is it correct that you see passing the `where` argument deeper into the creation of `RCfunctions`, and ultimately assigning that as the enclosing environment of the `ans` object which is created and returned, as fixing this scoping issue (for nimbleFunctions created in other packages)?
2. Furthermore, since this is a noteworthy change in the hierarchy of the enclosing environments of our functions, I also understand that we'd have to do some extensive testing / thinking about whether this change would "break" other things.

And frankly, regarding (2) above, I'd say it's almost likely that somewhere, somewhere in the entire code base, we rely somehow (in an ad-hoc way) on the current structure of the enclosing environments, for example to locate a particular object, or to test the type of a function or object.  Anyway, I'm just saying I think (2) above is a very legitimate concern, and if I had to guess, making this change **would** cause some sort of problem, elsewhere.

Likewise, I'm happy to discuss further here, or next week.

**Comment by paciorek** [comment]
1) Yes, I think this proposal would fix the scoping issue.
2) Point taken, though all the proposal would do is eliminate the two intervening enclosing environments for RCfunctions defined in the nimble package. I believe the remainder of the set of enclosing environments is the same.

**Comment by paciorek** [comment]
Ok, so on branch rc-scoping, I am fixing this. @danielturek I think you can start to proceed again with BayesNSGP, but note what I say next.

So far it seems to fix the issues with the BayesNSGP package checks with the exception that we still get
```
 * checking R code for possible problems ... NOTE
nimble_sparse_chol: no visible global function definition for
  ‘R_sparse_chol’
nimble_sparse_crossprod: no visible global function definition for
  ‘R_sparse_crossprod’
nimble_sparse_solve: no visible global function definition for
  ‘R_sparse_solve’
nimble_sparse_tcrossprod: no visible global function definition for
  ‘R_sparse_tcrossprod’
Undefined global functions or variables:
  R_sparse_chol R_sparse_crossprod R_sparse_solve R_sparse_tcrossprod
```

Note that the functions being flagged are all the output of `nimbleRcall`. So I think for both `nimbleRcall` and `nimbleExternalCall`, I need to look into scoping behavior there. 

@perrydv I see an `envir` arg to `nimbleRcall` (but not to `nimbleExternalCall`) but that arg is not used. So I think I have free reign to monkey with scoping here but let me know of any concerns/thoughts.

On first pass, it looks like I need to have a `where` arg (instead of `envir` for consistency) to `nimbleRcall` and pass that into `ans <- quote(RCfunction(fun, check = FALSE))` and similarly for `ans <- RCfunction(fun, check = FALSE)` in `nimbleExternalCall`.

**Comment by paciorek** [comment]
And a note just for potential future reference: 

nimbleFunctions (i.e., generators) created in a package that depends on nimble will have namespace:nimble as their enclosing environment.

```
parent.env(environment(BayesNSGP::calculateAD_ns1))
<environment: namespace:nimble>
```

nimbleFunction instances created in a package that depends on nimble will have namespace:other_package as their enclosing environment.

```
rfun = calculateAD_ns1()
parent.env(rfun)
<environment: namespace:BayesNSGP>
```

**Comment by danielturek** [comment]
Fantastic @paciorek .  I took a look at the changes, and it seems as we had discussed.

I also tried building / checking / installing the `BayesNSGP` package, using the updates you made on nimble branch `rc-scoping`, and just as you said, everything worked fine, except still for the warnings about "no visible global function definition" for the functions created using `nimbleRcall`.

So if I understood you, assuming @perrydv doesn't mind you also modifying the scoping for functions produced using `nimbleRcall`, then we can also resolve this through appropriate use of `where`.  Is that right?

Thanks for these fixes, @paciorek .
